src/System.Text.Json/src/System/Text/Json/JsonReaderState.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Defines an opaque type that holds and saves all the relevant state information which must be provided
+    /// to the <see cref="JsonUtf8Reader"/> to continue reading after processing incomplete data.
+    /// This type is required to support reentrancy when reading incomplete data, and to continue
+    /// reading once more data is available. Unlike the <see cref="JsonUtf8Reader"/>, which is a ref struct,
+    /// this type can survive across async/await boundaries and hence this type is required to provide
+    /// support for reading in more data asynchronously before continuing with a new instance of the <see cref="JsonUtf8Reader"/>.
+    /// </summary>
+    public struct JsonReaderState
+    {
+        // We are using a ulong to represent our nested state, so we can only
+        // go 64 levels deep without having to allocate a stack.
+        internal const int StackFreeMaxDepth = sizeof(ulong) * 8;
+
+        // This ulong container represents a tiny stack to track the state during nested transitions.
+        // The first bit represents the state of the current depth (1 == object, 0 == array).
+        // Each subsequent bit is the parent / containing type (object or array). Since this
+        // reader does a linear scan, we only need to keep a single path as we go through the data.
+        // This is primarily used as an optimization to avoid having to allocate a stack object for
+        // depths up to 64 (which is the default max depth).
+        internal ulong _stackFreeContainer;
+        internal long _lineNumber;
+        internal long _lineBytePosition;
```

Comment:
Do these two above need to be long?

Reply:
Yes, they do, to support `ReadOnlySequence`.